### PR TITLE
Feature/kotlin/library/retrofit

### PR DESCRIPTION
**Pull request by javils**
#4517

### PR checklist

- [X] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [X] If contributing template-only or documentation-only changes which will change sample output, [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) before.
- [X] Run the shell script(s) under `./bin/` (or Windows batch scripts under`.\bin\windows`) to update Petstore samples related to your fix. This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit, and these must match the expectations made by your contribution. You only need to run `./bin/{LANG}-petstore.sh`, `./bin/openapi3/{LANG}-petstore.sh` if updating the code or mustache templates for a language (`{LANG}`) (e.g. php, ruby, python, etc).
- [X] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master`, `4.3.x`, `5.0.x`. Default: `master`.
- [X] Copy the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) to review the pull request if your PR is targeting a particular programming language.

### PR Description
This PR add support for retrofit2 as a HTTP library in kotlin-client generator. The reason to add this feature is that I want to port some old android java code to kotlin, but I couldn't because retrofit2 was not supported.

@jimschubert  @dr4ke616 @karismann  @Zomzog  @andrewemery

### CODE REVIEW

File: bin/kotlin-client-all.sh [hunk 1]
```diff
@@ -8,3 +8,4 @@
 ./bin/kotlin-client-string.sh
 ./bin/kotlin-client-threetenbp.sh
 ./bin/kotlin-client-nullable.sh
+./bin/kotlin-client-retrofti2.sh
```
Comment (wing328) [inline bin/kotlin-client-all.sh]: typo: kotlin-client-retrofti2.sh => kotlin-client-retrofit2.sh

**Comment by wing328** [comment]
When using `--library retrofit2`, the following files are missing:
```
-rw-r--r--   1 wheel   4.0K Nov 22 13:46 README.md
-rw-r--r--   1 wheel   844B Nov 22 13:46 build.gradle
-rw-r--r--   1 wheel    35B Nov 22 13:46 settings.gradle
```

Can we also auto-generate these files similar to what we've done for other library options in the kotlin generator?

File: bin/kotlin-client-all.sh (hunk 1 above)
Comment (javils) [reply bin/kotlin-client-all.sh]: Thanks for reviewing! 
It's solved in this commit: a3b34b5

**Comment by javils** [comment]
> When using `--library retrofit2`, the following files are missing:
> 
> ```
> README.md
> build.gradle
> settings.gradle
> ```
> 
> Can we also auto-generate these files similar to what we've done for other library options in the kotlin generator?

Thanks! 
Solved in this commit: f37bfbb

**Comment by 4brunu** [comment]
@javils can you please run on the terminal `sh bin/utils/ensure-up-to-date --batch` and commit the file `docs/generators/kotlin.md` to make the CI pass?
Overall the PR looks ok, but it's a bit scary to merge this, because it changes so many things.

File: bin/kotlin-client-retrofit2.sh [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+  mvn -B clean package
+fi
+
+# if you've executed sbt assembly previously it will use that instead.
+export JAVA_OPTS="${JAVA_OPTS} -Xmx1024M -DloggerPath=conf/log4j.properties"
+ags="generate -t modules/openapi-generator/src/main/resources/kotlin-client -i modules/openapi-generator/src/test/resources/2_0/petstore.yaml -g kotlin --artifact-id kotlin-petstore-retrofit2 --library retrofit2 -o samples/client/petstore/kotlin-retrofit2 $@"
+
+java ${JAVA_OPTS} -jar ${executable} ${ags}
+
+cp CI/samples.ci/client/petstore/kotlin-threetenbp/pom.xml samples/client/petstore/kotlin-threetenbp/pom.xml
```
Comment (4brunu) [inline bin/kotlin-client-retrofit2.sh]: @javils this like is not related to this sample, can you please remove it?

Comment (javils) [reply bin/kotlin-client-retrofit2.sh]: Thanks for reviewing!

Fixed here: 6e4566f7

**Comment by javils** [comment]
> @javils can you please run on the terminal `sh bin/utils/ensure-up-to-date --batch` and commit the file `docs/generators/kotlin.md` to make the CI pass?
> Overall the PR looks ok, but it's a bit scary to merge this, because it changes so many things.

When I run this command I got a lot of modified files and not update the kotlin.md file. Any idea?

**Comment by 4brunu** [comment]
@javils Hey, I created this PR https://github.com/javils/openapi-generator/pull/1 to help you with the outdated files to make CI.

**Comment by javils** [comment]
@4brunu Thanks for the PR :-)

**Comment by javils** [comment]
@wing328 the build is failing because timeout in github repo, could you restart it?

**Comment by 4brunu** [comment]
@javils You can add an empty line, commit and push and the CI will restart

**Comment by wing328** [comment]
Restarted Drone.io job. Let's see how it goes.

Another way is to close and then re-open the PR (which will re-run ALL the CI jobs)

**Comment by javils** [comment]
Thanks!
For the next PR I know the trick.

**Comment by javils** [comment]
@wing328 @jimschubert @dr4ke616 @karismann @Zomzog @andrewemery
Can we merge this? Someone need to check something else?

**Comment by wing328** [comment]
I did a test with the auto-generated Kotlin code (retrofit2) for Petstore but got the following errors:
```
Starting a Gradle Daemon (subsequent builds will be faster)

> Task :compileKotlin
e: /private/tmp/re/src/main/kotlin/org/openapitools/client/apis/PetApi.kt: (18, 51): Unresolved reference: CSVParams
e: /private/tmp/re/src/main/kotlin/org/openapitools/client/apis/PetApi.kt: (20, 6): No value passed for parameter 'message'
e: /private/tmp/re/src/main/kotlin/org/openapitools/client/apis/PetApi.kt: (22, 45): Unresolved reference: CSVParams
e: /private/tmp/re/src/main/kotlin/org/openapitools/client/apis/PetApi.kt: (36, 130): Unresolved reference: MultipartBody
e: /private/tmp/re/src/main/kotlin/org/openapitools/client/apis/StoreApi.kt: (14, 22): Parameter name expected
e: /private/tmp/re/src/main/kotlin/org/openapitools/client/apis/StoreApi.kt: (14, 79): Expecting comma or ')'
e: /private/tmp/re/src/main/kotlin/org/openapitools/client/apis/StoreApi.kt: (16, 5): Expecting ')'
e: /private/tmp/re/src/main/kotlin/org/openapitools/client/apis/StoreApi.kt: (16, 6): Expecting member declaration
e: /private/tmp/re/src/main/kotlin/org/openapitools/client/apis/StoreApi.kt: (16, 9): Expecting member declaration
e: /private/tmp/re/src/main/kotlin/org/openapitools/client/apis/StoreApi.kt: (16, 10): Expecting member declaration
e: /private/tmp/re/src/main/kotlin/org/openapitools/client/apis/StoreApi.kt: (16, 11): Expecting member declaration
e: /private/tmp/re/src/main/kotlin/org/openapitools/client/apis/StoreApi.kt: (16, 33): Expecting member declaration
e: /private/tmp/re/src/main/kotlin/org/openapitools/client/apis/StoreApi.kt: (16, 34): Expecting member declaration
e: /private/tmp/re/src/main/kotlin/org/openapitools/client/apis/UserApi.kt: (29, 20): Parameter name expected
e: /private/tmp/re/src/main/kotlin/org/openapitools/client/apis/UserApi.kt: (29, 32): Expecting comma or ')'
e: /private/tmp/re/src/main/kotlin/org/openapitools/client/apis/UserApi.kt: (31, 5): Expecting ')'
e: /private/tmp/re/src/main/kotlin/org/openapitools/client/apis/UserApi.kt: (31, 6): Expecting member declaration
e: /private/tmp/re/src/main/kotlin/org/openapitools/client/apis/UserApi.kt: (31, 9): Expecting member declaration
e: /private/tmp/re/src/main/kotlin/org/openapitools/client/apis/UserApi.kt: (31, 10): Expecting member declaration
e: /private/tmp/re/src/main/kotlin/org/openapitools/client/apis/UserApi.kt: (31, 11): Expecting member declaration
e: /private/tmp/re/src/main/kotlin/org/openapitools/client/apis/UserApi.kt: (31, 27): Expecting member declaration
e: /private/tmp/re/src/main/kotlin/org/openapitools/client/apis/UserApi.kt: (31, 28): Expecting member declaration
e: /private/tmp/re/src/main/kotlin/org/openapitools/client/infrastructure/ApiClient.kt: (20, 30): Unresolved reference: MoshiConverterFactory

> Task :compileKotlin FAILED

FAILURE: Build failed with an exception.
```

Can you please take a look?

**Comment by javils** [comment]
> I did a test with the auto-generated Kotlin code (retrofit2) for Petstore but got the following errors:
> 
> ```
> Starting a Gradle Daemon (subsequent builds will be faster)
> 
> > Task :compileKotlin
> e: /private/tmp/re/src/main/kotlin/org/openapitools/client/apis/PetApi.kt: (18, 51): Unresolved reference: CSVParams
> e: /private/tmp/re/src/main/kotlin/org/openapitools/client/apis/PetApi.kt: (20, 6): No value passed for parameter 'message'
> e: /private/tmp/re/src/main/kotlin/org/openapitools/client/apis/PetApi.kt: (22, 45): Unresolved reference: CSVParams
> e: /private/tmp/re/src/main/kotlin/org/openapitools/client/apis/PetApi.kt: (36, 130): Unresolved reference: MultipartBody
> e: /private/tmp/re/src/main/kotlin/org/openapitools/client/apis/StoreApi.kt: (14, 22): Parameter name expected
> e: /private/tmp/re/src/main/kotlin/org/openapitools/client/apis/StoreApi.kt: (14, 79): Expecting comma or ')'
> e: /private/tmp/re/src/main/kotlin/org/openapitools/client/apis/StoreApi.kt: (16, 5): Expecting ')'
> e: /private/tmp/re/src/main/kotlin/org/openapitools/client/apis/StoreApi.kt: (16, 6): Expecting member declaration
> e: /private/tmp/re/src/main/kotlin/org/openapitools/client/apis/StoreApi.kt: (16, 9): Expecting member declaration
> e: /private/tmp/re/src/main/kotlin/org/openapitools/client/apis/StoreApi.kt: (16, 10): Expecting member declaration
> e: /private/tmp/re/src/main/kotlin/org/openapitools/client/apis/StoreApi.kt: (16, 11): Expecting member declaration
> e: /private/tmp/re/src/main/kotlin/org/openapitools/client/apis/StoreApi.kt: (16, 33): Expecting member declaration
> e: /private/tmp/re/src/main/kotlin/org/openapitools/client/apis/StoreApi.kt: (16, 34): Expecting member declaration
> e: /private/tmp/re/src/main/kotlin/org/openapitools/client/apis/UserApi.kt: (29, 20): Parameter name expected
> e: /private/tmp/re/src/main/kotlin/org/openapitools/client/apis/UserApi.kt: (29, 32): Expecting comma or ')'
> e: /private/tmp/re/src/main/kotlin/org/openapitools/client/apis/UserApi.kt: (31, 5): Expecting ')'
> e: /private/tmp/re/src/main/kotlin/org/openapitools/client/apis/UserApi.kt: (31, 6): Expecting member declaration
> e: /private/tmp/re/src/main/kotlin/org/openapitools/client/apis/UserApi.kt: (31, 9): Expecting member declaration
> e: /private/tmp/re/src/main/kotlin/org/openapitools/client/apis/UserApi.kt: (31, 10): Expecting member declaration
> e: /private/tmp/re/src/main/kotlin/org/openapitools/client/apis/UserApi.kt: (31, 11): Expecting member declaration
> e: /private/tmp/re/src/main/kotlin/org/openapitools/client/apis/UserApi.kt: (31, 27): Expecting member declaration
> e: /private/tmp/re/src/main/kotlin/org/openapitools/client/apis/UserApi.kt: (31, 28): Expecting member declaration
> e: /private/tmp/re/src/main/kotlin/org/openapitools/client/infrastructure/ApiClient.kt: (20, 30): Unresolved reference: MoshiConverterFactory
> 
> > Task :compileKotlin FAILED
> 
> FAILURE: Build failed with an exception.
> ```
> 
> Can you please take a look?

Thanks for your feedback!
I've done some commits to solve the issues :-)

**Comment by 4brunu** [comment]
@javils Since in your last commit you made a change in a `mustache`, you also need to update the pet projects, otherwise the CI will fail, because the pet projects are outdated.

**Comment by javils** [comment]
@4brunu done!

**Comment by javils** [comment]
The build failed 2 times because of this: 

> **An error occurred during the signature verification. The repository is not updated and the previous index files will be used. GPG error: http://deb.opera.com/opera stable InRelease: The following signatures couldn't be verified because the public key is not available: NO_PUBKEY**

**Comment by wing328** [comment]
@javils the shippable failure has been fixed in the master. Can you please merge the latest master into this branch to retrigger the build?

**Comment by wing328** [comment]
I still got errors when testing the Petstore samples (kotlin retrofit2):

```
> Task :compileKotlin
e: /Users/williamcheng/Code/openapi-generator/samples/client/petstore/kotlin-retrofit2/src/main/kotlin/org/openapitools/client/apis/PetApi.kt: (20, 6): No value passed for parameter 'message'
e: /Users/williamcheng/Code/openapi-generator/samples/client/petstore/kotlin-retrofit2/src/main/kotlin/org/openapitools/client/apis/PetApi.kt: (36, 130): Unresolved reference: MultipartBody
```

Does it work for you locally?

**Comment by 4brunu** [comment]
I have tested this locally and I think now it's working.
Can anyone confirm? 🙂

**Comment by wing328** [comment]
Tested locally and the result is good. Thanks for the PR.

**Comment by wing328** [comment]
UPDATE: the library option has been renamed from `retrofit2` to `jvm-retrofit2`
